[ALLUXIO-3301] S3A Streaming upload

LuQQiu opened the pull request:
https://alluxio.atlassian.net/browse/ALLUXIO-3301

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/229/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/255/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/256/
Test FAILed.

LuQQiu commented:
jenkins, test this please

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/258/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/260/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/279/
Test FAILed.

LuQQiu commented:
@madanadit @calvinjia PTAL, thanks!
Fulfill the main logic, now working on testing it and adding unit/integration tests.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/283/
Test FAILed.

calvinjia commented on core/client/fs/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java:
```diff
@@ -110,6 +113,9 @@
   private final Condition mBufferEmptyOrFailed = mLock.newCondition();
 
+  /** Flush should not be called before write. */
+  private boolean mWrote = false;
```
Could we allow the RPC to go through and no-op on the server side? This would simplify the client state.

LuQQiu replied:
Now the client side will send the flush command to worker side.
Worker side will not actually flush() if no write() occurred.

calvinjia commented on core/client/fs/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java:
```diff
@@ -481,13 +489,13 @@ public void operationComplete(ChannelFuture future) {
+   * The netty channel future listener that is called when a EOF or CANCEL or FLUSH is complete.
    */
-  private final class EofOrCancelListener implements ChannelFutureListener {
+  private final class EofOrCancelorFlushListener implements ChannelFutureListener {
```
This listener cannot be reused because `Cancel` and `EOF` are terminal RPCs whereas `flush` is not

LuQQiu replied:
Thanks for the kind remind. Add a new `FlushListener` for flush only.

calvinjia commented on integration/fuse/src/main/java/alluxio/fuse/AlluxioFuseFileSystem.java:
```diff
@@ -762,6 +771,30 @@ private int rmInternal(String path, boolean mustBeFile) {
+    } catch (TimeoutException te) {
+      return false;
+    }
+  }
```
If it is not too much trouble, could you put the FUSE changes in a separate PR? These seem like independent changes.

LuQQiu replied:
Sure, I would put Fuse changes in another PR, thanks!

calvinjia commented on core/common/src/main/java/alluxio/heartbeat/HeartbeatContext.java:
```diff
@@ -39,6 +39,8 @@
   public static final String MASTER_LOST_WORKER_DETECTION = "Master Lost Worker Detection";
   public static final String MASTER_METRICS_SYNC = "Master Metrics Sync";
+  public static final String MASTER_S3A_INTERMEDIATE_UPLOAD_CLEAN
+      = "Master S3A Intermediate Upload Clean";
```
I would suggest to make this agnostic to the UFS, ie. `Master UFS Cleanup` and call a `cleanup` method for each of the registered UnderFileSystems which for S3 does the existing logic. @madanadit what do you think about adding `cleanup` to the UFS interface?

LuQQiu replied:
Create a `UnderFileSystem.cleanup()` method. And setup periodically cleanup all non-readonly ufs in `DefaultFileSystemMaster`
Thanks

LuQQiu commented:
jenkins, test this please

calvinjia commented on core/server/master/src/main/java/alluxio/master/file/S3AUploadCleaner.java:
```diff
@@ -0,0 +1,146 @@
+ * S3A intermediate multipart upload periodic clean.
+ */
+@NotThreadSafe
+final class S3AUploadCleaner implements HeartbeatExecutor {
```
Would it be cleaner (no pun intended) to put this logic in the S3 under filesystem?

LuQQiu replied:
In the previous meeting, we thought it may be too many cleanup if we put it in the S3 under filesystem. But it would be definitely cleaner to put this logic in the S3AUnderFileSystem, we do not need to create additional transfer managers and could reuse nearly all of the codes.

LuQQiu replied:
Put the cleanup logic in the `S3AUnderFileSystem.cleanup`. Leader master will clean up all UFS periodically if `UNDERFS_CLEANUP_ENABLED`. This makes the code much cleaner! Thanks

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/284/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/288/
Test FAILed.

uronce-cc commented on underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java:
```diff
@@ -222,16 +233,21 @@ public static S3AUnderFileSystem createInstance(AlluxioURI uri,
+   * @param executor the executor for executing upload tasks
    * @param transferManager Transfer Manager for efficient I/O to S3
    * @param conf configuration for this S3A ufs
+   * @param streamingUploadEnabled whether streaming uplaod is enabled
```
NIT: typo, uplaod -> upload

LuQQiu replied:
Fixed, thanks

uronce-cc commented on underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java:
```diff
@@ -0,0 +1,406 @@
+ * transfer is done using an S3 low-level multipart upload API.
+ */
+@NotThreadSafe
+public class S3ALowLevelOutputStream extends OutputStream {
```
renamed to S3AStreamingOutputStream?

LuQQiu replied:
We first come up with the `S3AStreamingOutputStream` name and thinks it's a little strange. Haven't come up with better names....

uronce-cc commented on underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java:
```diff
@@ -0,0 +1,406 @@
+   * Only parts bigger than 5MB could be uploaded through multipart upload,
+   * except the last part.
+   */
+  private static final long UPLOAD_THRESHOLD = 5 * Constants.MB;
```
make this a configurable property in case users want to increase the threshold?

uronce-cc commented on underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java:
```diff
@@ -0,0 +1,406 @@
+    try {
+      mHash = MessageDigest.getInstance("MD5");
+    } catch (NoSuchAlgorithmException e) {
+      LOG.warn("Algorithm not available for MD5 hash.", e);
```
this should never happen, can throw RuntimeException

LuQQiu replied:
Agree, this should never happen, but throw an exception in the constructor should be avoided.
Putting mHash initialization in other methods looks wired, I suggest remaining it here.

uronce-cc commented on underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java:
```diff
@@ -0,0 +1,406 @@
+    // Partition size should be at least 5 MB, since S3 low-level multipart upload does not
+    // accept intermediate part smaller than 5 MB.
+    long partSize = Configuration.getBytes(PropertyKey.UNDERFS_S3A_STREAMING_UPLOAD_PARTITION_SIZE);
+    mPartitionSize = partSize < 5 * Constants.MB ? 5 * Constants.MB : partSize;
```
min(UPLOAD_THRESHOLD, partSize)

LuQQiu replied:
Changed, thanks!

uronce-cc commented on underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java:
```diff
@@ -0,0 +1,406 @@
+      meta.setSSEAlgorithm(ObjectMetadata.AES_256_SERVER_SIDE_ENCRYPTION);
+    }
+    if (mHash != null) {
+      meta.setContentMD5(new String(Base64.encode(mHash.digest())));
```
use `encodeAsString` in case the default system default charset decode the encoded bytes incorrectly in the constructor of `String`.

LuQQiu replied:
Thanks! Changed.

uronce-cc commented on underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java:
```diff
@@ -0,0 +1,406 @@
+  /**
+   * Executes the upload part request.
+   */
+  private void execUpload(UploadPartRequest request, File file, int partNumber) {
```
`file` and `partNumber` can be retrieved from `request`

LuQQiu replied:
Removed those parameters, thanks

uronce-cc commented on underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java:
```diff
@@ -0,0 +1,406 @@
+    // because Fuse release() method which calls close() is async.
+    // In flush(), we upload the current writing file if it is bigger than 5 MB,
+    // and wait for all current upload to complete.
+    mLocalOutputStream.flush();
```
If `flush` is called immediately after `write`, then `mLocalOutputStream` is `null`.

LuQQiu replied:
The mLocalOutputStream do have the chance to be null although it may be low.
Add a null check to make it more robust.
Changed to
```
    if (mLocalOutputStream != null) {
      mLocalOutputStream.flush();
    }
```
Thanks

uronce-cc commented on underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java:
```diff
@@ -0,0 +1,406 @@
+            .withPartNumber(partNumber)
+            .withFile(mFile)
+            .withPartSize(mFile.length());
+        uploadRequest.setLastPart(true);
```
if each time when write is called, length is exactly mPartitionSize, then when `close` is called, `mFile` is `null`, so `setLastPart(true)` will never be called, does `setLastPart` must be called for the last part?

uronce-cc commented on underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java:
```diff
@@ -0,0 +1,406 @@
+    } catch (ExecutionException e) {
+      // No recover ways so that we need to cancel all the upload tasks
+      // and abort the multipart upload
+      for (ListenableFuture<PartETag> future : mFutureTags) {
```
cancel the future returned by `addAll` will cancel all futures in mFutureTags.

LuQQiu replied:
Thanks! Great and convenient methods.

Changed to
```
Futures.allAsList(mFutureTags).cancel(true);
```

uronce-cc commented on underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java:
```diff
@@ -0,0 +1,406 @@
+      abortMultiPartUpload();
+      throw new IOException("Part upload failed in multipart upload with "
+          + "id '" + mUploadId + "' to " + mKey, e);
+    } catch (InterruptedException e) {
```
if any one of the threads in the executor is interrupted, `allAsList(...).get()` will throw this exception, but other threads in the executor may still be running, so futures need to be cancelled here too.

LuQQiu replied:
Make sense, if interrupted, we should also cancel all the uploads and abort the whole multipart upload. Thanks!
```
    } catch (InterruptedException e) {
      LOG.warn("Interrupted object upload.", e);
      Futures.allAsList(mFutureTags).cancel(true);
      abortMultiPartUpload();
      Thread.currentThread().interrupt();
    }
```

LuQQiu commented on underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java:
```diff
@@ -222,16 +233,21 @@ public static S3AUnderFileSystem createInstance(AlluxioURI uri,
     super(uri, conf);
     mClient = amazonS3Client;
     mBucketName = bucketName;
+    mExecutor = executor;
```
All the underFileSystem resources will be cleaned at the end when we 'cleanupRequest' or 'cancelRequest' in the 'UfsWriteHandler'.

LuQQiu commented on underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java:
```diff
@@ -0,0 +1,406 @@
+      } catch (AmazonClientException e) {
+        lastException = e;
+      }
+    } while (mRetryPolicy.attempt());
```
I want the whole upload request can retry for 5 times instead of each operation could retry for 5 times. If the whole request occurs errors more than 5 times, we better to cancel it and clean up.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/290/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/291/
Test FAILed.

LuQQiu commented:
@calvinjia @uronce-cc PTAL, thanks!
Main Changes:
(1) cleanup S3A intermediate upload in leader master by periodically calling `UnderFileSystem.cleanup`.
(2) removed Fuse open() delay to PR https://github.com/Alluxio/alluxio/pull/7807
(3) flush() will be sent to worker and flush() will not happens before write() in worker side.

LuQQiu commented:
@calvinjia The sync `flush()` trigger strange issue in `FileInStreamIntegrationTest` and I haven't find out the reason, so I roll back all the sync flush related changes in this PR. 
Make sync `flush()` another PR may be better to help find out the test failed reason. Thanks

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/300/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/328/
Test PASSed.

LuQQiu commented:
@calvinjia Added `S3ALowLevelOutputStreamTest` and fixed some bugs, PTAL, thanks

calvinjia commented on underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java:
```diff
@@ -0,0 +1,422 @@
+ * The data will be persisted to a temporary directory on the local disk
+ * and uploaded when the buffered data reaches the partition size. The data
+ * transfer is done using an S3 low-level multipart upload API.
+ */
```
Could you summarize the relationship between the methods and variables and the invariants? In particular, how the `init` methods manage the variables and the state of `mFile` and `mOffset` expected by each of the methods?

LuQQiu replied:
Provide the following detailed relationship and explanation to streaming upload.
```
/**
 * [Experimental] A stream for writing a file into S3 using streaming upload.
 * The data transfer is done using S3 low-level multipart upload.
 *
 * The multipart upload is initialized in the first write() and an upload id is given
 * by AWS S3 to distinguish different multipart uploads.
 *
 * We upload data in partitions. When write(), the data will be persisted to
 * a temporary file {@link #mFile} on the local disk. When the data {@link #mPartitionOffset}
 * in this temporary file reaches the {@link #mPartitionSize}, the file will be submitted
 * to the upload executor {@link #mExecutor} and we do not wait for uploads to finish.
 * A new temp file will be created for the future write and the {@link #mPartitionOffset}
 * will be reset to zero. The process goes until all the Fdata has been written to temp files.
 *
 * In flush(), we upload the buffered data if they are bigger than 5MB
 * and wait for all uploads to finish. The temp files will be deleted after uploading successfully.
 *
 * In close(), we upload the last part of data (if exists), wait for all uploads to finish,
 * and complete the multipart upload.
 *
 * close() will not be retried, but all the multipart upload
 * related operations(init, upload, complete, and abort) will be retried.
 *
 * If an error occurs and we have no way to recover, we abort the multipart uploads.
 * Some multipart uploads may not be completed/aborted in normal ways and need periodical cleanup
 * by enabling the {@link PropertyKey#UNDERFS_CLEANUP_ENABLED}.
 * When a leader master starts or a cleanup interval is reached, all the multipart uploads
 * older than {@link PropertyKey#UNDERFS_S3A_INTERMEDIATE_UPLOAD_CLEAN_AGE} will be cleaned.
 */
```
Any recommendation for the above docs here?

calvinjia commented on underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java:
```diff
@@ -181,10 +195,9 @@ public static S3AUnderFileSystem createInstance(AlluxioURI uri,
-        .setRequestTimeout((int) Configuration.getMs(PropertyKey.UNDERFS_S3A_REQUEST_TIMEOUT));
+    boolean streamingUploadEnabled =
+        conf.isSet(PropertyKey.UNDERFS_S3A_STREAMING_UPLOAD_ENABLED)
+        && conf.getBoolean(PropertyKey.UNDERFS_S3A_STREAMING_UPLOAD_ENABLED);
```
Do we need to check `isSet` since there is a default value?

LuQQiu replied:
I am a little confused with the configuration in `S3AUnderFileSystem`. Some configuration comes from the `UnderFileSystemConfiguration` and some come from the Global `Configuration`.

For here, `isSet` is not needed, I will remove it, thanks

calvinjia commented on underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java:
```diff
@@ -208,12 +221,13 @@ public static S3AUnderFileSystem createInstance(AlluxioURI uri,
+    TransferManager transferManager = TransferManagerBuilder.standard()
+        .withS3Client(amazonS3Client).withExecutorFactory(() -> service)
+        .withMultipartCopyThreshold(MULTIPART_COPY_THRESHOLD)
+        .build();
```
Is this the recommended way of constructing the transfer manager?

LuQQiu replied:
The old way has been deprecated so I change to use this way which the official doc recommends.

LuQQiu commented on core/common/src/main/java/alluxio/heartbeat/HeartbeatContext.java:
```diff
@@ -39,6 +39,8 @@
   public static final String MASTER_LOST_MASTER_DETECTION = "Master Lost Master Detection";
   public static final String MASTER_LOST_WORKER_DETECTION = "Master Lost Worker Detection";
   public static final String MASTER_METRICS_SYNC = "Master Metrics Sync";
+  public static final String MASTER_S3A_INTERMEDIATE_UPLOAD_CLEAN
```
Renamed to 
```
MASTER_UFS_CLEANUP = "Master Ufs Cleanup";
```
Thanks

LuQQiu commented on underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java:
```diff
@@ -0,0 +1,422 @@
+  private File mFile;
+
+  /** Store the future of tags. */
+  private List<ListenableFuture<PartETag>> mFutureTags = new ArrayList<>();
```
Renamed, thanks!

LuQQiu commented on underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java:
```diff
@@ -0,0 +1,422 @@
+  @Override
+  public void write(int b) throws IOException {
+    mSingleCharWrite[0] = (byte) b;
+    write(mSingleCharWrite, 0, 1);
```
Changed, thanks

LuQQiu commented on underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java:
```diff
@@ -0,0 +1,422 @@
+  @Override
+  public void write(byte[] b, int off, int len) throws IOException {
+    if (mUploadId == null) {
+      initMultiPartUpload();
```
Changed to
```
  @Override
  public void write(byte[] b, int off, int len) throws IOException {
    if (b == null || len == 0) {
      return;
    }
    validateWriteArgs(b, off, len);
    if (mUploadId == null) {
      initMultiPartUpload();
    }
    if (mFile == null) {
      initNewFile();
    }
```
do len and null check first, and then validate write arguments, see if any arguments out of range.
Finally init multipart upload or file.
Thanks

LuQQiu commented on underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java:
```diff
@@ -0,0 +1,422 @@
+  private OutputStream mLocalOutputStream;
+
+  /** When the offset reaches the partition size, we upload the temp file. */
+  private long mOffset;
```
Agree, changed to `mPartitionOffset` to improve the correctness.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/349/
Test PASSed.

LuQQiu commented:
@calvinjia Thanks for the suggestions. I have addressed the issues. PTAL, thanks!

LuQQiu commented:
The Streaming upload docs will be in another PR.

calvinjia commented:
@LuQQiu Do you know why the proto files are changed?

LuQQiu commented:
@calvinjia 
The proto files changed to include lineage deprecated information. Maybe because after we deprecate lineage, we do not regenerate the proto files. 

I first include sync flush in this PR and regenerate the proto files.
After that, I moved the sync flush to another PR, deleted the `WriteRequest.flush` in `protocol.proto` file and regenerate the proto files.

I think it is fine to include the proto file changes here since they only include lineage deprecate information.

LuQQiu commented:
I removed the proto files and let sync flush PR do the lineage deprecated changes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/353/
Test PASSed.
